internal/pkg/cli/deploy/workload.go

```diff
@@ -367,24 +392,86 @@ func (d *workloadDeployer) uploadContainerImages(out *UploadArtifactsOutput) err
 	if err != nil {
 		return fmt.Errorf("login to image repository: %w", err)
 	}
+
+	var mu sync.Mutex
 	out.ImageDigests = make(map[string]ContainerImageIdentifier, len(buildArgsPerContainer))
+
+	// counter for indexing the labeled output buffers.
+	bufferIdx := 0
+	// An array of buffers, one for each container image build and push output.
+	labeledBuffers := make([]*syncbuffer.LabeledSyncBuffer, len(buildArgsPerContainer))
+
+	// create a context and an error group.
+	g, ctx := errgroup.WithContext(context.Background())
+
+	// create a new cursor and hide it.
+	cursor := cursor.New()
+	cursor.Hide()
+
 	for name, buildArgs := range buildArgsPerContainer {
+		// create a copy of loop variables to avoid data race.
+		name := name
+		buildArgs := buildArgs
+
+		// create a pipe for streaming the build and push output.
+		pr, pw := io.Pipe()
+
 		buildArgs.URI = uri
+
+		// generate build args slice for docker build label.
 		buildArgsList, err := buildArgs.GenerateDockerBuildArgs(dockerengine.New(exec.NewCmd()))
 		if err != nil {
-			return fmt.Errorf("generate docker build args: %w", err)
-		}
-		// TODO(adi) handle this log in syncBuffer's Print function
-		log.Infof("Building your container image: docker %s\n", strings.Join(buildArgsList, " "))
-		digest, err := d.repository.BuildAndPush(buildArgs)
-		if err != nil {
-			return fmt.Errorf("build and push image: %w", err)
+			return fmt.Errorf("generate docker build args of image %s: %w", name, err)
 		}
-		out.ImageDigests[name] = ContainerImageIdentifier{
-			Digest:            digest,
-			CustomTag:         d.image.CustomTag,
-			GitShortCommitTag: d.image.GitShortCommitTag,
+		buf := syncbuffer.New()
+		labeledBuffer := buf.WithLabel(fmt.Sprintf("Building your container image %s: docker %s", name, strings.Join(buildArgsList, " ")))
+		labeledBuffers[bufferIdx] = labeledBuffer
+		bufferIdx++
+
+		g.Go(func() error {
+			defer pw.Close()
+			digest, err := d.repository.BuildAndPush(ctx, buildArgs, pw)
+			if err != nil {
+				return fmt.Errorf("build and push image %s: %w", name, err)
+			}
+			mu.Lock()
+			out.ImageDigests[name] = ContainerImageIdentifier{
+				Digest:            digest,
+				CustomTag:         d.image.CustomTag,
+				GitShortCommitTag: d.image.GitShortCommitTag,
+			}
+			mu.Unlock()
+			return nil
+		})
+		g.Go(func() error {
+			if err := buf.Copy(pr); err != nil {
+				return fmt.Errorf("copying build and push output for container %s: %w", name, err)
+			}
+			return nil
+		})
+	}
+	if isCIEnvironment() {
+		numLinesForBuildAndPush = -1
+	}
+	// create a LabeledTermPrinter for rendering build and push output.
+	ltp := d.labeledTermPrinter(os.Stderr, labeledBuffers, syncbuffer.WithNumLines(numLinesForBuildAndPush), syncbuffer.WithPadding(paddingForBuildAndPush))
+
+	g.Go(func() error {
+		for {
+			if ltp.IsDone() {
+				break
```

Comment:
once it's done, you might need one last `ltp.Print()` otherwise it's possible that it's over but we haven't printed our last view.

Reply:
@efekarakus I have a concern with the current logic of `printall`. 
let say we have 3 container images. In case of `CI=true` We are expecting to print the complete output of the images `only 3 times` one  time for each image. 
But with the current logic of `printall` let us say if one buffer is done and the remaining buffers takes some additional amount of time to be done. 

In this case, We are continuously printing the output of the first image buffer more than once until  all buffers are done.

So this is the reason i am thinking of having the `infinite for loop` with in the `syncbuffer` package and change the `printall to Public `  something like below
```
func (ltp *LabeledTermPrinter) PrintAll() {
doneCount := 0
	for {
		for idx, buf := range ltp.buffers {
			if !buf.syncBuf.IsDone() {
				continue
			}
			outputLogs := ltp.buffers[idx].syncBuf.strings()
			ltp.writeLines(buf.label, outputLogs)
			doneCount++
		}
		if doneCount >= len(ltp.buffers) {.  ----------> this way ensure we print only once from each buffer.
			break
		}
	}	
}
```	
With this approach I think I need to remove the logic of magic number `-1` for printing the entire content of the buffer.

Reply:
To address this concern
Modified `printall` function such that if the entire contents of a single buffer are printed, then remove that buffer from the buffers slice.